pkg/kubelet/kubelet.go

```diff
@@ -1025,137 +1014,204 @@ func (kl *Kubelet) syncPod(pod *api.BoundPod, dockerContainers dockertools.Docke
-				// If the container's liveness probe is unsuccessful, set readiness to false. If liveness is succesful, do a readiness check and set
-				// readiness accordingly. If the initalDelay since container creation on liveness probe has not passed the probe will return Success.
-				// If the initial delay on the readiness probe has not passed the probe will return Failure.
-				ready := probe.Unknown
-				live, err := kl.probeContainer(container.LivenessProbe, podFullName, uid, podStatus, container, dockerContainer, probe.Success)
-				if live == probe.Success {
-					ready, _ = kl.probeContainer(container.ReadinessProbe, podFullName, uid, podStatus, container, dockerContainer, probe.Failure)
-				}
-				if err != nil {
-					glog.V(1).Infof("liveness/readiness probe errored: %v", err)
-					containersToKeep[containerID] = empty{}
-					continue
-				}
-				if ready == probe.Success {
-					kl.readiness.set(dockerContainer.ID, true)
+			glog.Infof("pod %q container %q exists as %v", podFullName, container.Name, containerID)
+
+			if !createPodInfraContainer {
+				// look for changes in the container.
+				containerChanged := hash != 0 && hash != expectedHash
+				if !containerChanged {
+					// TODO: This should probably be separated out into a separate goroutine.
+					// If the container's liveness probe is unsuccessful, set readiness to false. If liveness is succesful, do a readiness check and set
+					// readiness accordingly. If the initalDelay since container creation on liveness probe has not passed the probe will return Success.
+					// If the initial delay on the readiness probe has not passed the probe will return Failure.
+					ready := probe.Unknown
+					live, err := kl.probeContainer(container.LivenessProbe, podFullName, uid, podStatus, container, dockerContainer, probe.Success)
+					if err != nil {
+						glog.Infof("liveness/readiness probe errored: %v", err)
+						containersToKeep[containerID] = &container
+						continue
+					}
+					if live == probe.Success {
+						ready, _ = kl.probeContainer(container.ReadinessProbe, podFullName, uid, podStatus, container, dockerContainer, probe.Failure)
+						kl.readiness.set(dockerContainer.ID, ready == probe.Success)
+						containersToKeep[containerID] = &container
+						continue
+					}
+					kl.readiness.set(dockerContainer.ID, ready == probe.Success)
+					ref, ok := kl.getRef(containerID)
+					if !ok {
+						glog.Warningf("No ref for pod '%v' - '%v'", containerID, container.Name)
+					} else {
+						record.Eventf(ref, "unhealthy", "Liveness Probe Failed %v - %v", containerID, container.Name)
+					}
+					glog.Infof("pod %q container %q is unhealthy (probe result: %v). Container will be killed and re-created.", podFullName, container.Name, live)
 				} else {
-					kl.readiness.set(dockerContainer.ID, false)
-				}
-				if live == probe.Success {
-					containersToKeep[containerID] = empty{}
-					continue
+					createPodInfraContainer = true
```

Comment:
Not really anything wrong, but: We don't necessarily need to do this. Today we do for a lot of cases we shouldn't and was a source of issues for the last bug in this code path. May be worth to change that as a separate PR after this.